url47e6aca666_pocorall_automaton_tgz-pJ8-net_pocorall_automaton_DatatypesJ8/src/net/pocorall/automaton/RunAutomatonMatcher.java

```diff
@@ -0,0 +1,299 @@
+/*
+ * dk.brics.automaton - AutomatonMatcher
+ *
+ * Copyright (c) 2008-2012 John Gibson
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. The name of the author may not be used to endorse or promote products
+ *    derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package net.pocorall.automaton;
+
+import java.util.regex.MatchResult;
+
+/**
+ * A tool that performs match operations on a given character sequence using
+ * a compiled automaton.
+ *
+ * @author John Gibson &lt;<a href="mailto:jgibson@mitre.org">jgibson@mitre.org</a>&gt;
+ * @see RunAutomaton#newMatcher(java.lang.CharSequence)
+ * @see RunAutomaton#newMatcher(java.lang.CharSequence, int, int)
+ */
+public class RunAutomatonMatcher implements MatchResult {
+
+	RunAutomatonMatcher(final CharSequence chars, final RunAutomaton automaton) {
+		this.chars = chars;
+		this.automaton = automaton;
+	}
+
+	private final RunAutomaton automaton;
+	private final CharSequence chars;
+
+	private int previousStart = -1;
+	private int matchStart = -1;
+
+	private int matchEnd = -1;
+
+	/**
+	 * Find the next matching subsequence of the input.
+	 * <br />
+	 * This also updates the values for the {@code start}, {@code end}, and
+	 * {@code group} methods.
+	 *
+	 * @return {@code not null} if there is a matching subsequence, null if there is no matching.
+	 */
+	public Object find() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.